drivers/dac/max538x/max538x.c

```diff
@@ -0,0 +1,188 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <errno.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include "max538x.h"
+#include "no_os_i2c.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+
+const struct max538x_chip_info chip_info[] = {
+	[MAX5380L] = {
+		.vfactor = 2,
+		.addr = 0x30,
+	},
+	[MAX5380M] = {
+		.vfactor = 2,
+		.addr = 0x31,
+	},
+	[MAX5380N] = {
+		.vfactor = 2,
+		.addr = 0x32,
+	},
+	[MAX5380K] = {
+		.vfactor = 2,
+		.addr = 0x33,
+	},
+	[MAX5381L] = {
+		.vfactor = 4,
+		.addr = 0x30,
+	},
+	[MAX5381M] = {
+		.vfactor = 4,
+		.addr = 0x31,
+	},
+	[MAX5381N] = {
+		.vfactor = 4,
+		.addr = 0x32,
+	},
+	[MAX5381K] = {
+		.vfactor = 4,
+		.addr = 0x33,
+	},
+	[MAX5382L] = {
+		.vfactor = 0.9,
+		.addr = 0x30,
+	},
+	[MAX5382M] = {
+		.vfactor = 0.9,
+		.addr = 0x31,
+	},
+	[MAX5382N] = {
+		.vfactor = 0.9,
+		.addr = 0x32,
+	},
+	[MAX5382K] = {
+		.vfactor = 0.9,
+		.addr = 0x33,
+	},
+};
+
+/***************************************************************************//**
+ * @brief Initializes the communication peripheral and checks if the device is
+ *        present.
+ *
+ * @param device     - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ *
+ * @return status - The result of the initialization procedure.
+ *                  Example: -1 - I2C peripheral was not initialized or the
+ *                                device is not present.
+ *                            0 - I2C peripheral was initialized and the
+ *                                device is present.
+*******************************************************************************/
+int32_t max538x_init(struct max538x_dev **device,
+		     struct max538x_init_param init_param)
+{
+	struct max538x_dev *dev;
+	int32_t status;
+
+	dev = (struct max538x_dev *)calloc(0, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	if ((init_param.max538x_vdd == NULL || init_param.max538x_vdd == 0)
+	    && dev->active_device > MAX5381K) {
+		free(dev);
+		return -EINVAL;
+	}
+
+	dev->active_device = init_param.active_device;
+	init_param.i2c_init.slave_address = chip_info[init_param.active_device].addr;
+
+	if (dev->active_device <= MAX5381K)
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor;
+	else
+		dev->max538x_vref = chip_info[init_param.active_device].vfactor *
+				    init_param.max538x_vdd;
+
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param.i2c_init);
+	if (status)
+		goto max538x_init_error;
+
+	*device = dev;
+
+	return 0;
+
+max538x_init_error:
+	max538x_remove(dev);
```

Comment:
best here would be to not use max538x_remove() but manually perform:
```
error_1:
    free(dev);
error:
    no_os_i2c_remove(dev->i2c_desc);
```

And jump to appropriate label with goto.